Added basic suite test

rhornung67 opened the pull request:
# Summary 

- This PR is an enhancement.
- It added a basic test to run in CI that runs all kernels and variants and verifies that they actually ran and that checksums are within 10e-7 tolerance of reference variant. We should be able to tighten this tolerance with a more robust checksum computation (see https://github.com/LLNL/RAJAPerf/pull/238). However, the test runs abbreviated versions of the kernels. The checksum diffs tend to get closer to zero as we run closer to the default.
- CI is enabled for azure and gitlab, although it could use some polishing.

kab163 commented:
> @kab163 i'll get to that after I change a few other things. Unfortunately, all recent xl versions have a bug in their openmp runtime, I believe, which causes all openmp variants of one kernel to generate garbage results

Uh Oh.. does that mean we need to exclude a specific subset of tests from running in the case that it's a xl+openmp job? (I'm thinking of how we did that for certain known-hip-failures for RAJA)

rhornung67 commented:
Probably.  I'll work on that after I get everything else working. Also, we found a bug in the RAJA CMake stuff that needs to be addressed before this PR can be merged. Working on that too.

rhornung67 commented on test/test-raja-perf-suite.cpp:
```diff
@@ -0,0 +1,120 @@
+
+// Assemble command line args for basic test
+  int argc = 4;
+#if defined(RAJA_ENABLE_HIP)
```
@MrBurmark here's what I came up with to disable the failing fused kernel test for hip builds. It's gross and should be checked with new versions. Any thoughts for improvement?

MrBurmark replied:
I think its reasonable for now, just have to add an issue to make it better later.
Would it be a bit simpler if you used sargv.size() instead of a separate argc?

MrBurmark replied:
Can you use a newer rocm version, kernel fusion appears to work here with rocm 5.1.

rhornung67 replied:
I used the argc/argv stuff from the start since it mimics the args to main, MPI_Init, etc. We would have to change the RunParams method that parses the command line arg list if we changed what we pass to the Executor ctor.

Of course, we can update the compiler versions. Let's do in a separate PR. Actually, that would be a good test to see if the RAJA dev guide CI docs I am working on are useful -- have someone see if they can do it following what's there.

rhornung67 replied:
@MrBurmark rocm 5.1 is available on rznevada. We're not quite there yet with CI testing for RAJA/RAJAPerf. @kab163 is working on it. On corona, the latest available is rocm 5.0.2.

MrBurmark replied:
We can use the HIP_VERSION to enable/disable the test so it automatically turns on when we can test it https://github.com/LLNL/RAJA/blob/develop/include/RAJA/config.hpp.in#L261.

rhornung67 replied:
So change the guard to this?

`#if (HIP_VERSION_MAJOR >= 5 && HIP_VERSION_MINOR >= 1)`

MrBurmark replied:
I think you want something more like this, or the negative of this actually in this case.

```
#if (HIP_VERSION_MAJOR > 5 || (HIP_VERSION_MAJOR == 5 && HIP_VERSION_MINOR >= 1))
```

rhornung67 replied:
I thought the test fails for rocm 5.0.2?

rhornung67 replied:
Gotcha.

MrBurmark commented on test/test-raja-perf-suite.cpp:
```diff
@@ -0,0 +1,120 @@
+
+  char** argv = new char* [argc];
+  for (int is = 0; is < argc; ++is) { 
+    argv[is] = const_cast<char*>(sargv[is].c_str());
```
Would it make more sense to make executor take const char** instead? We'd have to add a cast in main, but it could avoid const_cast here.

MrBurmark replied:
Or have two constructors for Executor to hide the cast of char** to const char** there.

rhornung67 replied:
I'm not sure it makes much difference. Again, following the common practice....

`int main( int argc, char** argv )`

MrBurmark replied:
Ya, const_cast is just evil.

rhornung67 replied:
I was trying to make the code simpler using C++ strings. c_str() returns const char*

MrBurmark replied:
Its fine, I was just hoping for an easy way to avoid the const_cast.

rhornung67 replied:
i kinda like it. it makes me feel naughty.

rhornung67 commented:
@MrBurmark This should be done now if you want to take a look.  Need to wait for PRs to be merged in RAJA (https://github.com/LLNL/RAJA/pull/1253) and radiuss-spack-configs (https://github.com/LLNL/radiuss-spack-configs/pull/31) and pull those in before we can merge this.

MrBurmark commented on test/test-raja-perf-suite.cpp:
```diff
@@ -0,0 +1,129 @@
+
+// Assemble command line args for basic test
+  int argc = 4;
+#if ( (defined(RAJA_ENABLE_HIP) && \
+       (HIP_VERSION_MAJOR < 5 || \
+       (HIP_VERSION_MAJOR == 5 && HIP_VERSION_MINOR < 1)) ) \
+      || (defined(RAJA_COMPILER_CLANG) && __clang_major__ == 11) )
+  argc = 6;
+#endif
```
I suppose these wouldn't overlap but should they still be separated to better match up with what is below?

rhornung67 replied:
Change pushed.

MrBurmark commented:
Looks pretty good, I'll take a final look after those PRs are in.

rhornung67 commented:
Folks, all PRs this depended on have been merged and submodules have been updated in this branch. So this should be good to merge when checks pass.  Please review. Thank you.

MrBurmark commented on .gitlab/lassen-jobs.yml:
```diff
@@ -26,18 +26,18 @@ clang_11_0_0:
 xl_16_1_1_11:
   variables:
-    SPEC: "%xl@16.1.1.11 cxxflags='-qthreaded -std=c++14 -O3 -qxlcompatmacros -qlanglvl=extended0x -qalias=noansi -qhot -qpic -qsmp=omp -qsuppress=1500-029 -qsuppress=1500-036'"
+    SPEC: "+openmp %xl@16.1.1.11 cxxflags='-qthreaded -std=c++14 -O3 -qstrict -qxlcompatmacros -qlanglvl=extended0x -qalias=noansi -qhot -qpic -qsmp=omp -qsuppress=1500-029 -qsuppress=1500-036'"
```
Should we document why we're using -qstrict?

rhornung67 replied:
I thought it would be batter to describe in an issue, so we can try to see if it works at some point.  Although, Mike C. this sort of thing has been an issue with XL for many years.

MrBurmark commented on CMakeLists.txt:
```diff
@@ -34,7 +31,20 @@ include(blt/SetupBLT.cmake)
 # Define RAJA PERFSUITE settings...
 #
 
+#cmake_dependent_option(RAJA_PERFSUITE_ENABLE_TESTS "Enable RAJA Perf Suite Tests" On "ENABLE_TESTS" Off)
+
+if (ENABLE_TESTS)
+
+  set(RAJA_PERFSUITE_ENABLE_TESTS On CACHE BOOL "Enable testing for RAJA Perf Suite")
```
Why is this done like this?

rhornung67 replied:
I wanted to turn off camp and RAJA tests because they take too long to build. I added a new variable to emphasize the point. It could be done other ways, I suppose. Is it confusing?

MrBurmark replied:
I mean why not use the commented out cmake_dependent_option?

rhornung67 replied:
Ah. I think I tried that and it didn't work for me. It may have been due to other breakage I was dealing with.

davidbeckingsale replied:
It would be good to use the dependent option if possible.

rhornung67 replied:
If I do that, I would have to set both ENABLE_TESTS and RAJA_PERFSUITE_ENABLE_TESTS to On? Or, am I mis-understanding?

davidbeckingsale replied:
The dependent option would set `RAJA_PERFSUITE_ENABLE_TESTS` to `On` if `ENABLE_TESTS` is On - so you only have to set `ENABLE_TESTS`.

rhornung67 replied:
Thanks. Will try.

davidbeckingsale commented on CMakeLists.txt:
```diff
@@ -34,7 +31,20 @@ include(blt/SetupBLT.cmake)
+
+endif()
+
+set(ENABLE_EXAMPLES Off CACHE BOOL "Enable RAJA examples")
```
Do we have a `RAJA_ENABLE_EXAMPLES` option that can be used instead?

davidbeckingsale replied:
Actually, looks like that is set below and this line can be deleted.

rhornung67 replied:
will do. thanks.

davidbeckingsale commented on test/CMakeLists.txt:
```diff
@@ -0,0 +1,25 @@
+# SPDX-License-Identifier: (BSD-3-Clause)
+###############################################################################
+
+include_directories(../src)
```
I think this should be a `target_include_directories` instead, e.g.:
```
target_include_directories(test-raja-perf-suite.exe PRIVATE ../src)
```

You could also use `PROJECT_SOURCE_DIR/src` to avoid using the relative path.

rhornung67 replied:
Using `PROJECT_SOURCE_DIR/src` the test doesn't build since it can't find the needed header files.  

`target_include_directories(test-raja-perf-suite.exe PRIVATE ../src)`  works if I put it after the `raja_add_test` macro call

rhornung67 replied:
Besides being more verbose, what is the advantage of this over what I had originally?

davidbeckingsale replied:
`${PROJECT_SOURCE_DIR}/src` should work. It's mostly a best practice thing.

rhornung67 replied:
change pushed to use target_include_directories

davidbeckingsale reviewed (dismissed):
This is fine just a couple of CMake comments

rhornung67 commented:
@MrBurmark @artv3 @davidbeckingsale @kab163 I've updated XL and fixed the build flags and updated to ROCm 5.1.0 for Corona. All checks should pass now, I've turned off allowing failures for XL+CUDA. Please review  and approve if you think this is good to merge.
